lib/SIL/Utils/GenericSpecializationMangler.cpp

```diff
@@ -0,0 +1,149 @@
+    DemangleInitRAII state(*this, MangledSpecialization, nullptr);
+    if (!parseAndPushNodes()) {
+      llvm::errs() << "Can't demangle: " << MangledSpecialization << '\n';
+      abort();
+    }
+    for (Node *Nd : NodeStack) {
+      addChild(Parent, Nd);
+    }
+  }
+};
+
+} // namespace
+
+std::string SpecializationMangler::finalize() {
+  StringRef MangledSpecialization(Storage.data(), Storage.size());
+  AttributeDemangler D;
+  NodePointer TopLevel = D.createNode(Node::Kind::Global);
+  D.demangleAndAddAsChildren(MangledSpecialization, TopLevel);
+
+  StringRef FuncName = Function ? Function->getName() : StringRef(FunctionName);
+  NodePointer FuncTopLevel = nullptr;
+  if (FuncName.startswith(MANGLING_PREFIX_STR)) {
+    FuncTopLevel = D.demangleSymbol(FuncName);
+    assert(FuncTopLevel);
+  }
+  if (!FuncTopLevel) {
+    FuncTopLevel = D.createNode(Node::Kind::Global);
+    FuncTopLevel->addChild(D.createNode(Node::Kind::Identifier, FuncName), D);
+  }
+  for (NodePointer FuncChild : *FuncTopLevel) {
+    TopLevel->addChild(FuncChild, D);
+  }
+  std::string mangledName = Demangle::mangleNode(TopLevel);
+  verify(mangledName);
+  return mangledName;
+}
+
+//===----------------------------------------------------------------------===//
+//                           Generic Specialization
+//===----------------------------------------------------------------------===//
+
+std::string GenericSpecializationMangler::mangle(GenericSignature Sig) {
+  beginMangling();
+
+  if (!Sig) {
+    assert(Function &&
+           "Need a SIL function if no generic signature is provided");
+    SILFunctionType *FTy = Function->getLoweredFunctionType();
+    Sig = FTy->getInvocationGenericSignature();
+  }
+
+  bool First = true;
+  Sig->forEachParam([&](GenericTypeParamType *ParamType, bool Canonical) {
+    if (Canonical) {
+      appendType(Type(ParamType).subst(SubMap)->getCanonicalType());
+      appendListSeparator(First);
+    }
+  });
+  assert(!First && "no generic substitutions");
+
+  if (isInlined)
+    appendSpecializationOperator("Ti");
+  else
+    appendSpecializationOperator(
+        isPrespecializaton ? "Ts" : (isReAbstracted ? "Tg" : "TG"));
+  return finalize();
+}
+
+static SubstitutionMap
+getSubstitutionMapForPrespecialization(GenericSignature genericSig,
+                                       GenericSignature specSig) {
+  auto CalleeGenericSig = genericSig;
+  auto SpecializedGenericSig = specSig;
+  auto SpecializedGenericEnv = specSig->getGenericEnvironment();
+  llvm::DenseMap<SubstitutableType *, Type>
+      CalleeInterfaceToSpecializedInterfaceMapping;
+
+  // TODO: Can this be simplified?
+  //
+
+  // Callee's generic signature and specialized generic signature
+  // use the same set of generic parameters, i.e. each generic
+  // parameter should be mapped to itself.
+  for (auto GP : CalleeGenericSig->getGenericParams()) {
+    CalleeInterfaceToSpecializedInterfaceMapping[GP] = Type(GP);
```

Comment:
Do you really need CalleeInterfaceToSpecializedInterfaceMapping at all then? The lookup() below can be replaced with just 'type' itself, no?

Reply:
Yes. Likely. I have copied this code from the FunctionSignaturePartialSpecializer and have not gone back to simplify it.